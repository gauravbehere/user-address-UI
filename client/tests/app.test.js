import React from 'react'
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import App from '../app.jsx';

describe('Test suite for App component', () => {

  const mockUserList = [{ name: 'Abc', empId: 123, gender: 'M', id: 2 }];
  const mockAddressList = [{text: 'Home 1', id: 1}];
  let fetchMock = null;
  const createStore = configureMockStore([thunk]);

  const getStore = (data) => {
    return createStore(data);
  }
  const respondWith = (obj, status) => {
    return () => Promise.resolve({
      json: () => Promise.resolve(obj),
      status
    });
  };

  beforeEach(() => {
    fetchMock = (obj, status) => {
      global.window.fetch = jest.fn().mockImplementation(respondWith(obj, status));
    }
  });

  it('Should render UserList component intially, fetching the users ', () => {
    const store = getStore({
      selelectedUserId: null,
      userList: mockUserList,
      addressList: [],
      userListError: null,
      addressListError: null
    });
    fetchMock(mockUserList, 200);
    const { container, getByText } = render(<Provider store={store}><App /></Provider>);
    expect(container).toMatchSnapshot();
    expect(getByText('Abc')).toBeInTheDocument();
  });

  it('Should render AddressList when there is a user selected', () => {
    const store = getStore({
      selelectedUserId: 1,
      userList: mockUserList,
      addressList: [{text: 'Home 1', id: 1}],
      userListError: null,
      addressListError: null
    });
    fetchMock(mockUserList, 200);
    const { getByText } = render(<Provider store={store}><App /></Provider>);
    expect(getByText('Home 1')).toBeInTheDocument();
  });


  it('Should fetch AddressList when a user is selected', () => {
    const store = getStore({
      selelectedUserId: 1,
      userList: mockUserList,
      addressList: [],
      userListError: null,
      addressListError: null
    });
    fetchMock(mockUserList, 200);
    const { getByTestId } = render(<Provider store={store}><App /></Provider>);
    fetchMock(mockAddressList, 200);
    fireEvent.click(getByTestId('user-list-item-0'));
    expect(fetch).toHaveBeenCalled();
  });

});