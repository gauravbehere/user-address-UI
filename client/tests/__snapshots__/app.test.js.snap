// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for App component Should render UserList component intially, fetching the users  1`] = `
<div>
  <div
    class="page-container"
  >
    <div
      class="user-list"
    >
      <div
        class="user-list-header"
      >
        List of Users
      </div>
      <div
        class="user-list-item"
        data-testid="user-list-item-0"
      >
        <div>
          Abc
        </div>
        <div>
          Age: 
        </div>
        <div>
          Employee#: 
          123
        </div>
        <div>
          Gender: 
          M
        </div>
      </div>
    </div>
    <div
      class="address-list"
    >
      <div
        class="address-list-header"
      >
        Address List
      </div>
      <div
        data-testid="noUserSelectedMsg"
      >
        Pick a user
      </div>
    </div>
  </div>
</div>
`;
